types/flatry/index.d.ts

```diff
@@ -3,7 +3,7 @@
 // Definitions by: po5epT <https://github.com/po5epT>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-declare function flatry<T>(promise: Promise<T>): Promise<[unknown] | [null, T]>;
-declare function flatry<T>(fn: () => T): [unknown] | [null, T];
+declare function flatry<T>(promise: Promise<T>): Promise<[unknown, never] | [null, T]>;
```

Comment:
This seems likely to be wrong - did you mean `null`, or `unknown`? A tuple containing `never` is semantically the same as `never`

Reply:
1. `unknown`, we don't know, how error would be thrown(mb not `Error` class), we try add second generic argument for this, but get error
![telegram-cloud-photo-size-2-5262725098710876298-y](https://user-images.githubusercontent.com/32272181/193517056-d83f631c-5d30-42e4-9789-4ca8ad452657.jpg)
2. In this case not, examples with current typing and old
```
const [err, res] = await flatry<string>(...);

if (err) {
  ...
}
// old
res.match() // error
// type of res is "string | undefined"

// new
res.match() // fine
// type of res is only "string"
```
Mb we need add it to tests, for clarity

Reply:
That only works because the union-ness of the result type is lost when you destructure it, and so `res` has type `T | never` which is just `T`; all you're doing is silencing the potentially undefined error.

It sounds to me like what you're trying to do is discriminate between the two potential tuples. It definitely won't work how you're trying to write it now as `unknown` can hold any value, including falsy ones, so your example code isn't guaranteed to work. It's certainly possible to write this code:

```ts
const [err, res] = flatry(() => throw false))
if (err) {
    // ...
} else {
    // ...
}
```

It seems like the only way to achieve what you're looking for is to leave the definitions as they are now and write:

```ts
const result = await flatry(...);

if (result.length === 1) {
    throw new Error("oops");
}

const x = result[1]
//    ^?
```

Which, reading the library, would work.

But, their examples show testing the error value instead, and I don't think there's a way to make that work besides using `!` to silence the error.